File: back-end/src/main/java/es/ull/project/domain/valueobject/Operation.java
package es.ull.project.domain.valueobject;

import es.ull.project.domain.enumerate.NodeContentType;
import es.ull.utils.docker.UllDockerImageName;

public class Operation extends NodeContent {

    private static final String ERROR_OPERATION_NAME_NOT_DEFINED = "Name for Operation not defined";
    private static final String ERROR_DOCKER_IMAGE_NOT_DEFINED = "Docker Image not defined";
    private final OperationName name;
    private final UllDockerImageName dockerImage;

    public Operation(OperationName name, UllDockerImageName dockerImage) {
        validateName(name);
        validateDockerImage(dockerImage);
        this.name = name;
        this.dockerImage = dockerImage;
    }

    private void validateName(OperationName name) {
        if (name == null) {
            throw new IllegalArgumentException(ERROR_OPERATION_NAME_NOT_DEFINED);
        }
    }

    private void validateDockerImage(UllDockerImageName dockerImage) {
        if (dockerImage == null) {
            throw new IllegalArgumentException(ERROR_DOCKER_IMAGE_NOT_DEFINED);
        }
    }

    public OperationName getName() {
        return this.name;
    }

    public UllDockerImageName getDockerImage() {
        return this.dockerImage;
    }

    public Operation setName(OperationName name) {
        return new Operation(name, this.dockerImage);
    }

    public Operation setDockerImage(UllDockerImageName dockerImage) {
        return new Operation(this.name, dockerImage);
    }

    @Override
    public NodeContentType getType() {
        return NodeContentType.OPERATION;
    }

    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (!(otherObject instanceof Operation)) {
            return false;
        }
        final Operation that = (Operation) otherObject;
        return this.name.equals(that.name) && this.dockerImage.equals(that.dockerImage);
    }

    @Override
    public int hashCode() {
        return 31 * name.hashCode() + dockerImage.hashCode();
    }

    @Override
    public String toString() {
        return String.format(
                "Operation={name=%s, dockerImage=%s}",
                this.name, this.dockerImage);
    }
}

File: front-end/src/domain/value-object/operation.ts
import { UllDockerImageName } from "@ull-tfg/ull-tfg-typescript/dist/ull-docker-image-name";
import { OperationName } from "./operation-name";
import { NodeContent } from "./node-content";
import { NodeContentType } from "../enumerate/node-content-type";


export class Operation extends NodeContent {
  private static readonly ERROR_OPERATION_NAME_NOT_DEFINED = 'Name for Operation not defined';
  private static readonly ERROR_DOCKER_IMAGE_NOT_DEFINED = 'Docker Image not defined';


  private readonly name: OperationName;
  private readonly dockerImage: UllDockerImageName;


  constructor(name: OperationName, dockerImage: UllDockerImageName) {
    super();
    this.validateName(name);
    this.validateDockerImage(dockerImage);
    this.name = name;
    this.dockerImage = dockerImage;

  }

  public getName(): OperationName {
    return this.name;
  }

  public getDockerImage(): UllDockerImageName {
    return this.dockerImage;
  }



  public setName(name: OperationName): Operation {
    return new Operation(name, this.dockerImage);
  }

  public setDockerImage(dockerImage: UllDockerImageName): Operation {
    return new Operation(this.name, dockerImage);
  }



  public getType(): NodeContentType {
    return NodeContentType.OPERATION;
  }

  private validateName(name: OperationName | null): void {
    if (!name) {
      throw new Error(Operation.ERROR_OPERATION_NAME_NOT_DEFINED);
    }
  }

  private validateDockerImage(dockerImage: UllDockerImageName | null): void {
    if (!dockerImage) {
      throw new Error(Operation.ERROR_DOCKER_IMAGE_NOT_DEFINED);
    }
  }



  public toString(): string {
    return `Operation{name=${this.name.toString()}, dockerImage=${this.dockerImage.toString()}}`;
  }
}

